linera-core/src/chain_worker/cache.rs

```diff
@@ -0,0 +1,259 @@
+            active_endpoints: Arc::new(Semaphore::new(limit.get())),
+        }
+    }
+
+    /// Obtains a [`ChainActorEndpoint`] to a [`ChainWorkerActor`] for a specific [`ChainId`], or
+    /// creates a new one if it doesn't yet exist.
+    pub async fn get_endpoint(
+        &self,
+        chain_id: ChainId,
+    ) -> Result<ChainActorEndpoint<StorageClient>, NewChainActorEndpoint<StorageClient>> {
+        let permit = self
+            .active_endpoints
+            .clone()
+            .acquire_owned()
+            .await
+            .expect("`active_endpoints` semaphore should never be closed");
+
+        self.try_get_existing_endpoint(chain_id, permit).map_err(
+            |MissingEndpointError {
+                 cache_is_full,
+                 permit,
+             }| {
+                if cache_is_full {
+                    self.stop_one();
+                }
+
+                self.create_new_endpoint(chain_id, permit)
+            },
+        )
+    }
+
+    /// Attempts to get a [`ChainActorEndpoint`] for a chain worker actor that's already running.
+    fn try_get_existing_endpoint(
+        &self,
+        chain_id: ChainId,
+        permit: OwnedSemaphorePermit,
+    ) -> Result<ChainActorEndpoint<StorageClient>, MissingEndpointError> {
+        let mut cache = self.cache.lock().unwrap();
+
+        if let Some(sender) = cache.get(&chain_id) {
+            Ok(ChainActorEndpoint::new(sender.clone(), permit))
+        } else {
+            Err(MissingEndpointError {
+                cache_is_full: cache.len() >= usize::from(cache.cap()),
+                permit,
+            })
+        }
+    }
+
+    /// Creates a new [`ChainActorEndpoint`], inserts it in the cache, and returns a
+    /// [`NewChainActorEndpoint`] ready to start the chain worker actor task.
+    fn create_new_endpoint(
+        &self,
+        chain_id: ChainId,
+        permit: OwnedSemaphorePermit,
+    ) -> NewChainActorEndpoint<StorageClient> {
+        let (sender, receiver) = mpsc::unbounded_channel();
+        self.cache.lock().unwrap().push(chain_id, sender.clone());
+
+        let delivery_notifier = self
+            .delivery_notifiers
+            .lock()
+            .unwrap()
+            .entry(chain_id)
+            .or_default()
+            .clone();
+
+        NewChainActorEndpoint {
+            chain_id,
+            delivery_notifier,
+            tasks: self.tasks.clone(),
+            sender: ChainActorEndpoint::new(sender, permit),
+            receiver,
+        }
+    }
+
+    /// Stops a single chain worker, opening up a slot for a new chain worker to be added.
+    fn stop_one(&self) {
+        let mut cache = self.cache.lock().unwrap();
+
+        let (&chain_to_evict, _) = cache
+            .iter()
+            .rev()
+            .find(|(_, candidate_endpoint)| candidate_endpoint.strong_count() <= 1)
+            .expect("`stop_one` should only be called while holding a permit for an endpoint");
```

Comment:
Can there be a race condition where the permit is already freed but the reference count hasn't been set to 0 yet?

Reply:
Good point! I was lucky that I declared the fields in `ChainActorEndpoint` in the right [order](https://doc.rust-lang.org/reference/destructors.html#destructors), so that the channel sender is always dropped before the permit :sweat_smile: